src/core/lib/security/transport/security_handshaker.c

```diff
@@ -223,17 +228,23 @@ static grpc_error *on_handshake_next_done_locked(
         GRPC_ERROR_CREATE_FROM_STATIC_STRING("Handshake failed"), result);
   }
   // Send data to peer.
-  grpc_slice to_send = grpc_slice_from_copied_buffer(
-      (const char *)bytes_to_send, bytes_to_send_size);
-  grpc_slice_buffer_reset_and_unref_internal(exec_ctx, &h->outgoing);
-  grpc_slice_buffer_add(&h->outgoing, to_send);
-  grpc_endpoint_write(exec_ctx, h->args->endpoint, &h->outgoing,
-                      &h->on_handshake_data_sent_to_peer);
-  // If handshake has completed, check peer and so on.
+  if (bytes_to_send_size > 0) {
+    grpc_slice to_send = grpc_slice_from_copied_buffer(
+        (const char *)bytes_to_send, bytes_to_send_size);
+    grpc_slice_buffer_reset_and_unref_internal(exec_ctx, &h->outgoing);
+    grpc_slice_buffer_add(&h->outgoing, to_send);
+    grpc_endpoint_write(exec_ctx, h->args->endpoint, &h->outgoing,
+                        &h->on_handshake_data_sent_to_peer);
+  }
+  // If handshake has completed, check peer and so on. Otherwise, need to read
+  // more data from the peer.
   if (handshaker_result != NULL) {
     GPR_ASSERT(h->handshaker_result == NULL);
     h->handshaker_result = handshaker_result;
     error = check_peer_locked(exec_ctx, h);
+  } else {
+    grpc_endpoint_read(exec_ctx, h->args->endpoint, h->args->read_buffer,
```

Comment:
Thinking about this further, I'm not sure it's a good idea to have both a read and a write outstanding at the same time.  The problem is that this requires more complex error-handling code.  For example, let's say that we're doing a write to the peer and expecting to read a response back.  If the write fails, then we still have a pending callback from the read, and we need to avoid destroying the handshaker until both callbacks have been invoked.  We could probably make that work with some refcounting changes, but it would be fairly complex, and I think we can avoid it by making this simpler.

I think we need to make two changes:

1. Put an `else` at the start of line 241 above, so that the three possible actions here (write, read, and handshake done) are mutually exclusive -- we can only choose one of them in any given call to this function.

2. Restore the write code in `on_handshake_data_sent_to_peer()`, and make it unconditional (i.e., it should not depend on whether `h->handshaker_result` is set).